plugins/wporg-5ftf/includes/contributor.php

```diff
@@ -566,3 +583,199 @@ function parse_contributors( $contributors, $pledge_id = null ) {
+		}
+
+		if ( ! $on_notifiable_team ) {
+			unset( $xprofiles[ $index ] );
+			continue;
+		}
+	}
+
+	return $xprofiles;
+}
+
+/**
+ * Merge user data with xprofile data.
+ */
+function add_user_data_to_xprofile( array $xprofiles ) : array {
+	global $wpdb;
+
+	if ( empty( $xprofiles ) ) {
+		return array();
+	}
+
+	$full_users      = array();
+	$xprofiles       = array_column( $xprofiles, null, 'user_id' ); // Re-index for direct access.
+	$user_ids        = array_keys( $xprofiles );
+	$id_placeholders = implode( ', ', array_fill( 0, count( $user_ids ), '%d' ) );
+
+	// Get user data.
+	// Ignore new users because they haven't had a chance to contribute yet.
+	// phpcs:disable -- `$id_placeholders` is safely created above.
+	$established_users = $wpdb->get_results( $wpdb->prepare( "
+		SELECT
+		    u.ID, u.user_email, u.user_login, u.user_nicename,
+			GROUP_CONCAT( um.meta_key ) AS meta_keys,
+			GROUP_CONCAT( um.meta_value ) AS meta_values
+		FROM `$wpdb->users` u
+			JOIN `$wpdb->usermeta` um ON u.ID = um.user_id
+		WHERE
+			um.user_id IN ( $id_placeholders ) AND
+			um.meta_key IN ( 'last_logged_in', '5ftf_last_inactivity_email', 'first_name' ) AND
+			u.user_registered < CURDATE() - INTERVAL %d MONTH
+		GROUP BY um.user_id
+		ORDER BY u.ID",
+		array_merge( $user_ids, array( INACTIVITY_THRESHOLD_MONTHS ) )
+	) );
+	// phpcs:enable
+
+	foreach ( $established_users as $user ) {
+		$full_user = array(
+			'user_id'        => absint( $user->ID ),
+			'user_email'     => $user->user_email,
+			'hours_per_week' => $xprofiles[ $user->ID ]->hours_per_week,
+			'user_nicename'  => $user->user_nicename,
+		);
+
+		$keys   = explode( ',', $user->meta_keys );
+		$values = explode( ',', $user->meta_values );
+
+		foreach ( $keys as $index => $key ) {
+			$full_user[ $key ] = maybe_unserialize( $values[ $index ] );
+		}
+
+		$full_user['last_logged_in']             = intval( strtotime( $full_user['last_logged_in'] ?? '' ) ); // Convert `false` to `0`.
+		$full_user['5ftf_last_inactivity_email'] = intval( $full_user['5ftf_last_inactivity_email'] ?? 0 );
+		$full_user['teams_names']                = (array) maybe_unserialize( $xprofiles[ $user->ID ]->team_names );
+
+		$full_users[] = $full_user;
+	}
+
+	return $full_users;
+}
+
+/**
+ * Prune users who shouldn't be notified of their inactivity.
+ */
+function prune_unnotifiable_users( array $contributors ) : array {
+	$inactivity_threshold = strtotime( INACTIVITY_THRESHOLD_MONTHS . ' months ago' );
+
+	foreach ( $contributors as $index => $contributor ) {
+		if ( $contributor['last_logged_in'] > $inactivity_threshold ) {
+			unset( $contributors[ $index ] );
+		}
+
+		if ( $contributor['5ftf_last_inactivity_email'] > $inactivity_threshold ) {
+			unset( $contributors[ $index ] );
+		}
+	}
+
+	return $contributors;
+}
+
+/**
+ * Notify an inactive contributor.
+ */
+function notify_inactive_contributor( array $contributor ) : void {
+	if ( ! Email\send_contributor_inactive_email( $contributor ) ) {
```

Comment:
🤔 that's true. 

It looks like account registration and editing both validate the address with Core's `is_email()`. That's not RFC-compliant, but that might actually help us in this case, since IIRC there are lots of "valid" addresses that aren't routable. There's probably some edge cases though.

`wp_mail()` also uses `is_email()` to validate. So, I think an address will always either: 
1. be blocked by `wp_mail()` and therefore not trigger any spam filters that hurt our SMTP IP reputation; or
2. be sent to the MTA and therefore get an updated `5ftf_last_inactivity_email`

In the case of `1`, though, that could maybe eventually cause some users to never get the email. If enough edge cases clumped together inside a batch group, then they could get stuck at the front of it, causing the others at the end of the group to always be ignored.

https://github.com/WordPress/five-for-the-future/blob/8d6ea36d2fd998b4b18bd178198e83cd3f0141c5/plugins/wporg-5ftf/includes/contributor.php#L596-L599

Adding a `shuffle()` in there would avoid that.

Reply:
I updated 041abcf to include `shuffle()`